### ISSUE TITLE
Multiscale layers have a shift due to handling of center vs corner of pixel/voxel

### ISSUE THREAD

**Issue by psobolewskiPhD** [napari/napari#6320]
## 🐛 Bug

While working on tests for https://github.com/napari/napari/pull/6317 with @brisvag we noticed that there's an offset that occurs with multiscale images/labels because of the difference between the *corner* of the pixel/voxel and the center. When rescaling using the center, the corner will not be fixed (and vice versa).
Worse yet, this effect occurs whenever `scale` is used and becomes noticeable when scale values approach the same order of magnitude as dimensions of a layer.

## To Reproduce

This script recapitulates the issue using multiscale labels--just toggle between 3D and 2D:
[Note: on main `get_value` will return an error, which is fixed in #6317 but this can be ignored here)

```
import napari
import numpy as np

def instantiate_3D_multiscale_labels():
    lowest_res_scale = np.arange(8).reshape(2, 2, 2)
    middle_res_scale = (
        lowest_res_scale.repeat(2, axis=0).repeat(2, axis=1).repeat(2, axis=2)
    )
    highest_res_scale = (
        middle_res_scale.repeat(2, axis=0).repeat(2, axis=1).repeat(2, axis=2)
    )

    return [highest_res_scale, middle_res_scale, lowest_res_scale]

viewer = napari.Viewer()
multiscale_labels = instantiate_3D_multiscale_labels()
points = np.array([[0, 0, 0], [0, 1, 1], [0, 2, 2]])

viewer.add_labels(multiscale_labels)
viewer.add_points(points)
viewer.dims.current_step = (0, 1, 1)
viewer.dims.ndisplay = 3

napari.run()
```

In **2D** the *highest resolution* is used for the labels. 
In **3D** the *lowest* resolution is used. 
You can see the labels shift relative to the points when you toggle back and forth.

## Expected behavior

There is good rationale for using corner or center of pixel/voxel. This came up in conversation at the hackathon that maybe this should be something that could be controlled by the user. 
But in this case of multiscale image/label the downsampled data should exactly overlay the original, with no shift, so this needs to be taken into account for scaling.

## Environment

```
❯ napari --info               
napari: 0.5.0a2.dev443+gc9da57b1
Platform: macOS-13.5.1-arm64-arm-64bit
System: MacOS 13.5.1
Python: 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]
Qt: 6.5.2
PyQt6: 
NumPy: 1.25.2
SciPy: 1.11.1
Dask: 2023.8.0
VisPy: 0.13.0
magicgui: 0.7.2
superqt: unknown
in-n-out: 0.1.8
app-model: 0.2.0
npe2: 0.7.2

OpenGL:
  - GL version:  2.1 Metal - 83.1
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 2056x1329, scale 2.0

Settings path:
  - /Users/sobolp/Library/Application Support/napari/napari-dev_36963bf0846bfe99dcf4f566a2cc780b6a723afa/settings.yaml
Plugins:
  - napari: 0.5.0a2.dev261+ge62c9dd3 (77 contributions)
  - napari-animation: 0.0.8.dev2+ga9ca7a7.d20230817 (2 contributions)
  - napari-console: 0.0.8 (0 contributions)
  - napari-svg: 0.1.10 (2 contributions)

```

**Comment by jni** [comment]
> But in this case of multiscale image/label the downsampled data should exactly overlay the original, with no shift, so this needs to be taken into account for scaling.

I'm not sure but I don't think this is *necessarily* correct. It all depends on how the multiscale downsampling was generated, but you can imagine a process by which the lower resolution images don't exactly overlay the higher ones (e.g. they were acquired independently with some jitter, or the higher resolution only covers a certain region).

The issue here is #3783 and to a certain extent https://github.com/napari/napari/pull/3715#issuecomment-998127947. Basically, we need an explicit model for the transformations *at each level of the pyramid*. To move things along, sure, we can just *guess* the translation based on the constraint you've described. I basically did this here:

https://gist.github.com/jni/0377bd93b6be1b75348c6ca4ef2001c6#file-write_ome-py-L34-L41

even though napari doesn't support different transformations for different resolution levels. But we can do the same hardcoding for now and move to support explicit transforms for each level in the future, based on the work ome-ngff is doing.

PS: I *don't* think we should allow corner-centered coordinates. Even the name is an oxymoron. 😜

**Comment by brisvag** [comment]
Yeah, definitely a patch solution. But it's definitely better than what we have now for most cases :)

**Comment by andy-sweet** [comment]
@psobolewskiPhD : can you try re-explaining the `scale` issue? and/or provide any reproducers for that?

The reproducer in the description uses the default scale of (1, 1, 1) and then relies on changes to the multi-scale resolution due to the way the associated transforms are handled for 2D and 3D slices (which is why #6319 fixes the reproducer), but may not fix all underlying issues.

**Comment by psobolewskiPhD** [comment]
Hmm, I'm not 100% sure, this was something we discussed at the hackathon on the whiteboard. 
I will check notes on my work computer.
I think it may be as simple as:
Open cells3D
`viewer.layers[0].scale = (1, 5, 5)`
and zoom into corner of the image:
<img width="974" alt="image" src="https://github.com/napari/napari/assets/76622105/3adee300-e9d2-4827-9d5b-049aa8d392b0">

**Comment by andy-sweet** [comment]
Thanks for the quick follow up!

> Open cells3D
> `viewer.layers[0].scale = (1, 5, 5)`
> and zoom into corner of the image:

I'd argue that example is working as intended. If you independently specify two different scales for images with the same offset/translate and we want the first pixel to be centered on `(0, 0)` for both of them, then this is as expected.

For multi-scale (the way it is in napari now), we can add an offset to account for the scale of each level (assuming some simple downsampling procedure as napari does) according [to the diagram I made](https://github.com/napari/napari/pull/6319#discussion_r1363143858). That's what #6319 adds explicitly for 3D multi-scale images, whereas 2D multi-scale images are handled slightly different.

I get the sense that both 2D and 3D fail in complex cases (e.g. using `scale` and `translate` with multi-scale images, but I'm still struggling to write meaningful test cases -  which is why I asked.

**Comment by psobolewskiPhD** [comment]
Relevant to the discussion on center vs corner:
Question from Eugene Katrukha
> I was wondering if there is a convention for the voxel center in 3D volume rendering.
Let's say voxel (0,0,0), should it be rendered spanning from -0.5 to 0.5 or from 0 to 1 (kinda like ImageJ)?

https://imagesc.zulipchat.com/#narrow/channel/327238-Fiji/topic/Voxel.20position.20with.20respect.20to.20coordinates.20origin

Answer from John Bogovic:
> I like the first - so does BigDataViewer. The sample is at (0,0), and the pixel/voxel covers [-0.5, 0.5] because that what nearest-neighbor interpolation does. It's also the convention that ITK and the medical imaging folks have been using for a long time (I often point people to [Fig 4.1 of the ITK software guide](https://itk.org/ItkSoftwareGuide.pdf))

https://imagesc.zulipchat.com/#narrow/channel/327238-Fiji/topic/Voxel.20position.20with.20respect.20to.20coordinates.20origin/near/508300242

**Comment by brisvag** [comment]
Note: The example in the issue description is outdated, because with #6319 we actually fixed the multiscale offset which was necessary for the example code to show the problem.

Now, the offset only exists between *different* layers with different scales.
